pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java

```diff
@@ -0,0 +1,156 @@
+ *
+ * <p>It provide the higher level entry interface to convert a SQL string into a {@link QueryPlan}.
+ */
+public class QueryEnvironment {
+  private final RelDataTypeFactory _typeFactory;
+  private final CalciteSchema _rootSchema;
+  private final WorkerManager _workerManager;
+  private final FrameworkConfig _config;
+
+  private final PlannerImpl _planner;
+  private final SqlValidator _validator;
+  private final Prepare.CatalogReader _catalogReader;
+  private final Collection<RelOptRule> _logicalRuleSet;
+  private final RelOptPlanner _relOptPlanner;
+
+  public QueryEnvironment(TypeFactory typeFactory, CalciteSchema rootSchema, WorkerManager workerManager) {
+    _typeFactory = typeFactory;
+    _rootSchema = rootSchema;
+    _workerManager = workerManager;
+    _config = Frameworks.newConfigBuilder().traitDefs().build();
+
+    // this is only here as a placeholder for SqlToRelConverter expandView implementation.
+    _planner = new PlannerImpl(_config);
+
+    // catalog
+    _catalogReader = new CalciteCatalogReader(_rootSchema, _rootSchema.path(null), _typeFactory, null);
+    _validator = new Validator(SqlStdOperatorTable.instance(), _catalogReader, _typeFactory);
+
+    // optimizer rules
+    _logicalRuleSet = PinotQueryRuleSets.LOGICAL_OPT_RULES;
+
+    // optimizer
+    HepProgramBuilder hepProgramBuilder = new HepProgramBuilder();
+    for (RelOptRule relOptRule : _logicalRuleSet) {
+      hepProgramBuilder.addRuleInstance(relOptRule);
+    }
+    _relOptPlanner = new LogicalPlanner(hepProgramBuilder.build(), Contexts.EMPTY_CONTEXT);
+  }
+
+  public QueryPlan sqlQuery(String sqlQuery) {
+    PlannerContext plannerContext = new PlannerContext();
+    try {
+      SqlNode parsed = this.parse(sqlQuery, plannerContext);
+      SqlNode validated = this.validate(parsed);
+      RelRoot relation = this.toRelation(validated, plannerContext);
+      RelNode optimized = this.optimize(relation, plannerContext);
+      return this.toQuery(optimized, plannerContext);
+    } catch (Exception e) {
+      throw new RuntimeException("Error composing query plan", e);
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // steps
+  // --------------------------------------------------------------------------
+
+  protected SqlNode parse(String query, PlannerContext plannerContext)
+      throws Exception {
+    // 1. invoke CalciteSqlParser to parse out SqlNode;
+    SqlNode compiled = CalciteSqlParser.compile(query, plannerContext);
+    // 2. TODO: add query rewrite logic
+    return compiled;
+  }
+
+  protected SqlNode validate(SqlNode parsed)
+      throws Exception {
+    // 3. validator to validate.
+    SqlNode validated = _validator.validate(parsed);
+    if (null == validated || !validated.getKind().belongsTo(SqlKind.QUERY)) {
+      throw new IllegalArgumentException(
+          String.format("unsupported SQL query, cannot validate out a valid sql from:\n%s", parsed));
```

Comment:
Can we include the original SQL query as well ?

Reply:
original query will be wrapped in upper-level planQuery

Reply:
Basically I wanted to include the original string SQL query in string. May be SqlNode.toString() takes care of that ?

Reply:
yes. one will not directly call validate. when it goes throw planQuery it will print the original query string.